### ISSUE TITLE
Checkpoint usability issues

### ISSUE THREAD

**Issue by siriuslee** [mosaicml/composer#414]
There are a number of usability issues that have surfaced while documenting checkpoints. 

- [x] `CheckpointLoader`, `CheckpointSaver` and `RunDirectoryUploader` all have pretty out of date docstrings (Pain level: moderate)
- [ ] `save_folder` that is a relative path should be relative to `CWD`, not the run directory. (Pain level: high)
- [ ] Saving to object store requires the `RunDirectoryUploader` callback to be provided as well. This is not obvious and probably shouldn't be required (Pain level: mild)
- [x] Checkpoints should be saved as pytorch checkpoint files, not tarballs (Pain level: high)

Additional notes:
**Save folder**
Since the user is specifying the `save_folder`, it really should be respected. If they used a relative path, assume they meant relative to the CWD, not relative to some run directory. If we are concerned about checkpoint files overwriting others, we can either error or use the timestamp to ensure the filename is unique. 

**Saving to object stores**
This is not particularly high priority, but is a usability issue. To use the object store, we should do as much as possible to just specify the URI as the `save_folder` and not require a separate callback.

**Comment by ravi-mosaicml** [comment]
> There are a number of usability issues that have surfaced while documenting checkpoints.
> 
> * [ ]  `CheckpointLoader`, `CheckpointSaver` and `RunDirectoryUploader` all have pretty out of date docstrings (Pain level: moderate)
Agreed; this will be fixed by #396 
> * [ ]  `save_folder` that is a relative path should be relative to `CWD`, not the run directory. (Pain level: high)
OK, this can be changed.
> * [ ]  Saving to object store requires the `RunDirectoryUploader` callback to be provided as well. This is not obvious and probably shouldn't be required (Pain level: mild)
For this one and the above one, we should probably remove the run directory, and instead just have an object store provider pool that is attached directly to the state. Callbacks that currently write to the run directory would write directly to the object store provider. (A local directory could also function as an object store provider) 
> * [ ]  Checkpoints should be saved as pytorch checkpoint files, not tarballs (Pain level: high)
Could you elaborate on the use case for accessing a checkpoint directly? I don't think there is a standard format for how to checkpoint the training state.
> 
> Additional notes: **Save folder** Since the user is specifying the `save_folder`, it really should be respected. If they used a relative path, assume they meant relative to the CWD, not relative to some run directory. If we are concerned about checkpoint files overwriting others, we can either error or use the timestamp to ensure the filename is unique.
> 
> **Saving to object stores** This is not particularly high priority, but is a usability issue. To use the object store, we should do as much as possible to just specify the URI as the `save_folder` and not require a separate callback.
You would need credentials to upload to a URI, in which case you would need the object store provider.

**Comment by hanlint** [comment]
Good points. For checkpointing, ideally our format would work with `torch.load` and `torch.save`, and use the `*.pt` extension. The use case for loading the checkpoint directly is, say, analyzing the model weights, or even extracting them for deployment purposes with a different toolchain.

**Comment by hanlint** [comment]
Perhaps we try to hit up the "High" pain level issues in v0.4, and defer the rest of v0.5?

**Comment by ravi-mosaicml** [comment]
Agree it would make sense to have a weights only checkpoint format for these cases. Is there any industry standard? (Onnx?)

**Comment by siriuslee** [comment]
I think the main point is to make the file loadable with `torch.load`. It's fine if the `model` weights are nested, ie `state_dict["state"]["model"]`. Just using a tarball is going to look strange to any user of Pytorch.

**Comment by siriuslee** [comment]
> I presume one would need credentials to upload to an object store, in which case it would be more than a URI.

You're definitely right. I think it should basically mirror the args from `CheckpointLoader` with an `ObjectStoreProvider`.

**Comment by ravi-mosaicml** [comment]
>  save_folder that is a relative path should be relative to CWD, not the run directory. (Pain level: high)
I realize that this will be a bit more complicated, since we will need to differentiate between relative paths that are relative to the run directory vs to CWD. Can we defer this to 0.5 (thinking to fix it with something like #421) and just make it super clear in the docstring to use an absolute path if you don't want it relative to the run directory? Alternatively we could add a `relative_to_run_directory` flag as a short-term solution.

**Comment by hanlint** [comment]
I'll add another one in (cc: @ajaysaini725 

- [ ] We want to programmatically retrieve the paths of the saved checkpoints after training. e.g.:
```
trainer.fit()

saved_checkpoints = trainer.checkpoint_saver.saved_files  # absolute path
```

**Comment by ravi-mosaicml** [comment]
#522 fixed the checkpointing docstrings, and #470 fixed the callback docstrings

**Comment by ravi-mosaicml** [comment]
#421 and the new checkpoint API (https://www.notion.so/Checkpoint-API-Redesign-f0ee88d4bd5149b6a091589ed3f465f5) will fix the other issues identified in this issue. However, these changes are scheduled for v0.5, not v0.4.1
